### ISSUE TITLE
mouseEmulation vs. head tracking

### ISSUE THREAD

**Issue by luxuchenghu** [cybereality/Perception#49]
I am very interested in your great work and start reading this code recently. 
Here is my question, 
void MotionTracker::updateOrientation()
{
...
	if (mouseEmulation)
		SendInput(1, &mouseData, sizeof(INPUT));
	}
...
}
I found it is mouse emulation, but not head tracking. It uses delta yaw and delta pitch working as a mouse. Have anyone thought about mapping the head position into those normal 3D games? I mean we still send mouse signal while head moving, and monitorinig  pEye, pAt of the "D3DXMatrixLookAtLH" function, until it meet the head position.

**Comment by baggyg** [comment]
Because of VR Boost none of the current official list actually use mouse emulation. Instead we "push" the readings from the head tracker into the memory locations responsible for orientation. This has massive benefits in latency and compatibility with gamepads. 

I am not entirely sure of what you are suggesting but if you are talking about moving the Matrix due to headtracking we have tried something similar and it leads to lots of problems and artifacts as the game engine has culled items outside of the player's view.

**Comment by luxuchenghu** [comment]
Hi baggyg!
Thank you for your kind reply.
You mentioned, " Instead we ‘push' the readings from the head tracker into the memory locations responsible for orientation."
So where is " the memory locations responsible for orientation", how could we get the memory locations from games? Would you please find me the codes, in which file?
I only saw
...
if (mouseEmulation)
SendInput(1, &mouseData, sizeof(INPUT));
...
in line 135, file MotionTracker.cpp

**Comment by baggyg** [comment]
In the main perception all that happens is the vrBoost values are set. VR Boost is a separate DLL which is close source. 

We find the locations via Cheat Engine and determine a pattern (where they usually appear and what values - then V Boost scans the memory to find them).

**Comment by luxuchenghu** [comment]
"We find the locations via Cheat Engine and determine a pattern ."
So would you please find me the codes, in which file, and line from where to where? I am very curious about how does it work.

**Comment by baggyg** [comment]
Cheat Engine is a separate piece of software. You can find tutorials and a download on the site: http://www.cheatengine.org/

As I said, VR Boost is closed source so unfortunately I won't be able to show you how this gets applied, but should be pretty easy to see what it does once you have Cheat Engine working.

**Comment by luxuchenghu** [comment]
I am playing "Mount&Blade" using VIREIO, I have done following things before I start the game.
1, copied "d3d9.dll", "hijackdll.dll", "libfreespace.dll", "VRboost.dll" to the game folder;
2, Add in text below to /cfg/profiles.xml;
<profile game_name="mb_warband" game_exe="mb_warband.exe" game_type="0" />
3, Choosed side by side and my head tracker.
That's all.
Finally I found both my head tracker and the mouse can control the orientation in the game.

In my mind, if I were wearing my head tracker and looking up, no matter how I move my mouse, what I saw should be "looking up", but the mouse can turn the view down even if I was looking up.
So, does this was what expected to be?

How could I use VRboost, is "VRboostReferee.exe" can be used, should I generate a xml file for a game just like those in "/cfg/VRboost_rules"?

What is the relationship between "VRboost.dll" and "cheatengine", is "VRboost.dll" programmed referencing "cheatengine"?
